referrer-policy/no-referrer-policy/no-referrer-policy.html

```diff
@@ -0,0 +1,132 @@
+        iframe.src = url;
+        document.body.appendChild(iframe);
+
+        return iframe;
+      }
+
+      // Creates a test arrangement and appends an iframe to the document.
+      // Note: the URL gets concatenated with the ID of the arrangement.
+      function arrangeReferrerPolicyAsyncTest(id, url, description) {
+        var iframe = appendIframeToBody(url + "?id=" + id);
+        var test = async_test(description);
+        var testArrangement = {iframe:iframe, test:test};
+
+        return testArrangement;
+      }
+
+      // Asserts the values for an iframe response message and matched test
+      // arrangement.
+      function assertChildMessage(message, testArrangement) {
+        testArrangement.test.step(function() {
+              // Defense: Making sure the test itself is not broken.
+              assert_own_property(message, "id");
+              assert_own_property(message, "location");
+              assert_own_property(message, "referrer");
+              assert_own_property(testArrangement, "test");
+              assert_own_property(testArrangement, "iframe");
+
+              // Sanity check - location of iframe matches reported location.
+              assert_equals(message.location, testArrangement.iframe.src,
+                            "Child reported location.");
+
+              // Child reports the referrer URL - main test assertion.
+              assert_equals(message.referrer, expectedReferrerUrl,
+                            "Child reported referrer.");
+
+              // Also check the request headers reported from the server.
+              assert_equals(message.headers.referer, expectedReferrerUrl,
+                            "Child reported referrer (from server).");
+
+            },
+            "Referrer URL is passed to the sub-resource (iframe)."
+        );
+
+        testArrangement.test.done();
+      }
+
+      // Using a map to easily find and run the test corresponding to the
+      // received message ID (iframe).
+      var testArrangementMap = {};
+
+      // With this handler we receive all the messages from the iframes.
+      // We identify the source and appropriate test arrangement to be run via
+      // the provided message ID.
+      window.addEventListener("message", function (event) {
+        var childMessage = event.data;
+        var testArrangement = testArrangementMap[childMessage.id];
+        // Run the test for this sub-resource.
+        assertChildMessage(childMessage, testArrangement);
+      });
+
+      // Map all test arrangements by ID. With this mapping loop we also append
+      // the iframes to the body.
+      for (var i in scenarios) {
+        var scenario = scenarios[i];
+        testArrangementMap[scenario.id] =
+            arrangeReferrerPolicyAsyncTest(scenario.id,
+                                           scenario.url,
+                                           scenario.description);
+      }
```

Comment:
I think we'll end up wanting to move some substantial subset of this code out into generic methods in the common JS file as you add more and more tests. There's no reason to do that right now, but keep it in mind for future patches.